### PR TITLE
Tests refactoring

### PR DESCRIPTION
**Pull request by FrNecas**

### CODE REVIEW

File: rebasehelper/tests/test_specfile.py [hunk 1]
```diff
@@ -79,20 +81,43 @@ def test_set_version(self, spec_object):
...
-    def test__write_spec_content(self, spec_object):
+    @pytest.mark.parametrize('spec_attributes, content', [
+        (
+            {
+                'path': SPEC_FILE,
+            },
+            'testing line1\ntesting line 2\n'
+        )
+    ])
+    def test__write_spec_content(self, mocked_spec_object, content):
```
Comment (nforro) [inline rebasehelper/tests/test_specfile.py]: This makes me wonder - why do we test this? I mean, what does it actually test? `SpecContent.__str__()` or `file.write()`? I don't think the latter needs to be tested, and the former is already being tested by `TestSpecContent.test_string_representation()`.

File: rebasehelper/tests/test_specfile.py [hunk 2]
```diff
@@ -79,20 +81,43 @@ def test_set_version(self, spec_object):
...
         # pylint: disable=protected-access
-        new_content = 'testing line 1\ntesting line 2\n'
-        spec_object.spec_content = SpecContent(new_content)
-        spec_object._write_spec_content()
-        with open(SPEC_FILE) as spec:
-            assert new_content == spec.read()
+        mocked_spec_object.spec_content = SpecContent(content)
```
Comment (nforro) [inline rebasehelper/tests/test_specfile.py]: Why is this not a `spec_content` attribute inside `spec_attributes`?

File: rebasehelper/tests/test_specfile.py [hunk 3]
```diff
@@ -115,11 +140,6 @@ def test_get_patches(self, workdir, spec_object):
             patches[index] = [p.path, p.index]
         assert patches == expected_patches
 
-    def test_get_requires(self, spec_object):
-        expected = {'openssl-devel', 'pkgconfig', 'texinfo', 'gettext', 'autoconf'}
-        req = spec_object.header.requires
-        assert len(expected.intersection(req)) == len(expected)
-
```
Comment (nforro) [inline rebasehelper/tests/test_specfile.py]: I agree that this test makes no sense, however, there is nothing about removal of meaningless/redundant tests in the commit message ;-)

File: rebasehelper/tests/test_specfile.py (hunk 1 above)
Comment (FrNecas) [reply rebasehelper/tests/test_specfile.py]: I think that the original intention was to check whether replacing the whole `SpecContent` has an effect on writing the spec. You are right that there is probably no point in testing a method which just writes a string representation of `SpecContent` (which is already covered by tests). I will remove the test.

File: rebasehelper/tests/test_specfile.py [hunk 4]
```diff
@@ -177,23 +243,50 @@ def test_set_extra_version(self, spec_object):
...
+    def test_update_setup_dirname(self, mocked_spec_object):
+        # mock %{version} and %{name} macros, set their level to -3
+        macros = {'name': 'test', 'version': '1.0.2'}
+        for macro, value in macros.items():
+            rpm.addMacro(macro, value)
+        mocked_spec_object.macros = MacroHelper.dump()
+        for m in macros:
+            for macro in mocked_spec_object.macros:
+                if macro['name'] == m:
+                    macro['level'] = -3
```
Comment (nforro) [inline rebasehelper/tests/test_specfile.py]: This definitely shouldn't be part of a test. `mocked_spec_object` fixture should instead accept another argument like `macros`.

File: rebasehelper/tests/test_specfile.py [hunk 5]
```diff
@@ -177,23 +243,50 @@ def test_set_extra_version(self, spec_object):
...
         assert '-n %{name}-%{version}rc1' in setup
 
-        spec_object.update_setup_dirname('test-1.0.2-rc1')
-        prep = spec_object.spec_content.section('%prep')
+        mocked_spec_object.update_setup_dirname('test-1.0.2-rc1')
+        prep = mocked_spec_object.spec_content.section('%prep')
         setup = [l for l in prep if l.startswith('%setup')][0]
         assert '-n %{name}-%{version}-rc1' in setup
 
+        rpm.reloadConfig()
```
Comment (nforro) [inline rebasehelper/tests/test_specfile.py]: Same as above. This should be the *teardown* part of `mocked_spec_object` fixture.

File: rebasehelper/tests/test_specfile.py (hunk 4 above)
Comment (nforro) [reply rebasehelper/tests/test_specfile.py]: Or maybe, to stay consistent, it should process `macros` key in `spec_attributes` analogically to how `spec_content` is being processed.

File: rebasehelper/tests/plugins/test_spec_hooks.py [hunk 6]
```diff
@@ -70,28 +112,59 @@ def test_escape_macros_spec_hook(self, spec_object):
...
+
+                %changelog
+                * Wed Apr 26 2017 Nikola Forró <nforro@redhat.com> - 1.0.2-34
+                - Update to 1.0.2
+                """),
+            'header': SimpleNamespace(version='1.0.2'),
+        }
+    ])
+    def test_replace_old_version_spec_hook(self, mocked_spec_object, replace_with_macro):
+        new_spec = copy.deepcopy(mocked_spec_object)
```
Comment (nforro) [inline rebasehelper/tests/plugins/test_spec_hooks.py]: How about another fixture, called e.g. `mocked_spec_object_copy`, for this?

File: rebasehelper/tests/plugins/test_spec_hooks.py [hunk 7]
```diff
@@ -70,28 +112,59 @@ def test_escape_macros_spec_hook(self, spec_object):
...
         # Check if the version has been updated
-        test_source = [line for line in new_spec.spec_content.section('%package') if line.startswith('Source9')]
+        test_source = new_spec.get_raw_tag_value('Source9')
         assert test_source
         if replace_with_macro:
             expected_result = 'https://test.com/#/1.1/%{name}-hardcoded-version-%{version}.tar.gz'
         else:
             expected_result = 'https://test.com/#/1.1/%{name}-hardcoded-version-1.1.0.tar.gz'
-        assert test_source[0].split()[1] == expected_result
+        assert test_source == expected_result
```
Comment (nforro) [inline rebasehelper/tests/plugins/test_spec_hooks.py]: Now it doesn't make much sense not to assert directly in the condition. The change also makes `assert test_source` redundant.

File: rebasehelper/tests/plugins/test_spec_hooks.py [hunk 8]
```diff
@@ -70,28 +112,59 @@ def test_escape_macros_spec_hook(self, spec_object):
...
+    @pytest.mark.parametrize('spec_attributes', [
+        {
+            'spec_content': dedent("""\
+                URL: https://pypi.python.org/pypi/%{name}
+                Source0: https://pypi.python.org/.../%{name}.%{version}.tar.gz
+                """)
+        }
+    ])
+    def test_pypi_to_python_hosted_url_trans(self, mocked_spec_object):
         # pylint: disable=protected-access
```
Comment (nforro) [inline rebasehelper/tests/plugins/test_spec_hooks.py]: This is no longer necessary, is it?

File: rebasehelper/tests/conftest.py [hunk 9]
```diff
@@ -70,21 +71,26 @@ def mocked_spec_object(spec_attributes):
...
+            for macro, properties in value.items():
+                rpm.addMacro(macro, properties.get('value', ''))
+            macros = MacroHelper.dump()
+            for macro, properties in value.items():
+                for m in macros:
+                    if m['name'] == macro:
+                        for prop, v in properties.items():
+                            if prop != 'value':
+                                m[prop] = v
+            value = macros
```
Comment (nforro) [inline rebasehelper/tests/conftest.py]: Why this move? I mean, why not do it this way already in commit 0e7473d?

File: rebasehelper/tests/conftest.py [hunk 10]
```diff
@@ -70,21 +71,26 @@ def mocked_spec_object(spec_attributes):
...
-                        if prop != 'value':
-                            m[prop] = value
     if hasattr(spec, 'spec_content') and not hasattr(spec, 'tags'):
         spec.tags = Tags(spec.spec_content, spec.spec_content)
     return spec
+
+
+@pytest.fixture
+def mocked_spec_object_copy(mocked_spec_object):  # pylint: disable=redefined-outer-name
+    return copy.deepcopy(mocked_spec_object)
```
Comment (nforro) [inline rebasehelper/tests/conftest.py]: This is a nitpick, but still. Why are these new fixtures below `pytest_collection_modifyitems`  hook and not below `spec_object` fixture?

File: rebasehelper/tests/plugins/test_spec_hooks.py [hunk 11]
```diff
@@ -33,12 +36,38 @@
...
+                /usr/share/man/man1/*
+                /usr/bin/%{name}
+                %config(noreplace) /etc/test/test.conf
+
+                %files devel
+                %{_bindir}/test_example
+                %{_libdir}/my_test.so
+                /usr/share/test1.txt
+                /no/macros/here
+
```
Comment (nforro) [inline rebasehelper/tests/plugins/test_spec_hooks.py]: Why the empty line here?

File: rebasehelper/tests/plugins/test_spec_hooks.py [hunk 12]
```diff
@@ -33,12 +36,38 @@
...
+                """),
+            'macros': {},
```
Comment (nforro) [inline rebasehelper/tests/plugins/test_spec_hooks.py]: Why is this necessary?

File: rebasehelper/tests/conftest.py (hunk 9 above)
Comment (FrNecas) [reply rebasehelper/tests/conftest.py]: Hmm, I amended to a wrong commit :/ I didn't really like setting the `macros` attribute to spec (which aren't really the macros intended for use) and wanted to modify the attribute correctly inside the `for` loop, to stay consistent with how `SpecContent` is handled.

File: rebasehelper/tests/plugins/test_spec_hooks.py (hunk 12 above)
Comment (FrNecas) [reply rebasehelper/tests/plugins/test_spec_hooks.py]: The hook requires `SpecFile.macros` to be set. `mocked_spec_content` only sets the macros if `macros` key is a part of `spec_attributes`. By setting it to empy dict, `SpecFile.macros` is set to `MacroHelper.dump()`

File: rebasehelper/tests/test_specfile.py [hunk 13]
```diff
@@ -170,20 +232,49 @@ def test_set_extra_version(self, spec_object):
...
+                        {
+                            'value': 'test',
+                            'level': -3,
+                        },
+                    'version':
+                        {
+                            'value': '1.0.2',
+                            'level': -3,
+                        },
+                },
```
Comment (nforro) [inline rebasehelper/tests/test_specfile.py]: I think this would actually be more readable if it was squashed into 2 lines.

File: rebasehelper/tests/conftest.py [hunk 14]
```diff
@@ -53,24 +55,36 @@ def spec_object(workdir):  # pylint: disable=redefined-outer-name
...
+                for m in macros:
+                    if m['name'] == macro:
+                        for prop, v in properties.items():
+                            if prop != 'value':
+                                m[prop] = v
+            value = macros
         if attribute == 'spec_content' and isinstance(value, str):
             value = SpecContent(value)
         setattr(spec, attribute, value)
+
```
Comment (nforro) [inline rebasehelper/tests/conftest.py]: Extra newline :-)
